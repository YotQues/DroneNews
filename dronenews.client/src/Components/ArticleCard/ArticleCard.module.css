.card {
    border: 1px solid var(--borderColor);
    border-radius: 5px;
    height: 200px;
    display: flex;
    gap: 12px;
    transition: all .15s ease;
}

.card:hover {
    cursor: pointer;
    box-shadow: 8px 4px 20px rgba(0, 0 ,0, 0.2);
}

.card .image {
    flex: 1 0 160px;
    height: 100%;
    background-position: center;
}

.card .image img {
    object-fit: cover;
    max-width: 100%;
    height: 100%;
}

.card .text {
    padding: 2px;
}

.card .title {
}

.card .desc {
}

.card .content {
}
