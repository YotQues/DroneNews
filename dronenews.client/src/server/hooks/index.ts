export * from './useArticles.ts';
export { getSearchParams } from '../../utils/getSearchParams.ts';