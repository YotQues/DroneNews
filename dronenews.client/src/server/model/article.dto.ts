import type {SourceDto} from "./source.dto.ts";

export interface ArticleDto {
    id: number;
    title: string;
    description: string;
    content: string;
    originalUrl: string;
    imageUrl?: string;
    publishedAt: Date;
    author?: string;
    source?: SourceDto;
}

