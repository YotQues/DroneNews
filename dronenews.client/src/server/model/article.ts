import type {Source} from "./source.dto.ts";

export interface Article {
    id: number;
    title: string;
    description: string;
    content: string;
    originalUrl: string;
    imageUrl?: string;
    publishedAt: string;
    author?: string;
    source?: Source;
}

