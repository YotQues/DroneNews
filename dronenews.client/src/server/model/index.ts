export * from "./author.dto.ts"
export * from "./source.dto.ts"
export * from "./article.dto.ts"
export * from "./list-response.dto.ts"
