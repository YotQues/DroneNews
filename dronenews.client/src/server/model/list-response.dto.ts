export interface ListResponse<TData> {
    totalItems: number;
    items: TData[];
    isLastPage: boolean;
}

