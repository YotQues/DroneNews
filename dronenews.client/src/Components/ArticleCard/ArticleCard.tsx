import { Dialog, DialogContent, DialogTitle, Typography } from '@mui/material';
import { useState } from 'react';
import imgPlaceholder from '../../assets/image-placeholder.png';
import { Article } from '../../server/model';
import styles from './ArticleCard.module.css';

class IArticleCardProps {
  article: Article;
}

export function ArticleCard({ article }: IArticleCardProps) {
  const [imageError, setImageError] = useState(false);
  const [isDialogOpen, setIsDialogOpen] = useState(false);
  return (
    <div onClick={() => setIsDialogOpen(true)} className={styles.card + ' gridItem'}>
      <div className={styles.image}>
        <img
          onError={() => setImageError(true)}
          src={article.imageUrl == null || imageError ? (imgPlaceholder as string) : article.imageUrl}
          alt={'article-image ' + article.title}
        />
      </div>
      <div className={styles.text}>
        <div>
          <Typography variant="subtitle1" className={styles.title}>
            {article.title}
          </Typography>
          {article.publishedAt && <Typography variant="subtitle2">{article.publishedAt.split('T')[0]}</Typography>}
        </div>
      </div>
      <Dialog open={isDialogOpen} onClose={() => setIsDialogOpen(false)} >
        <DialogTitle>{article.title}</DialogTitle>

        <DialogContent>
          <Typography variant="h6">{article.description}</Typography>
          <Typography variant="p">{article.content}</Typography>
        </DialogContent>
      </Dialog>
    </div>
  );
}
