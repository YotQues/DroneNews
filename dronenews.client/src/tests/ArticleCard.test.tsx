import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent, act } from '@testing-library/react';
import { Article } from '../server/model';
import { ArticleCard } from '../Components/ArticleCard/ArticleCard.tsx';

const mockArticle: Article = {
  id: 1,
  title: 'Test Article',
  description: 'This is a test description.',
  content: 'This is the test content.',
  imageUrl: 'test-image-url',
  publishedAt: '2024-07-24T12:34:56Z',
} as any;

describe('ArticleCard', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<ArticleCard article={mockArticle} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display the article title and published date', () => {
    const { getByText } = render(<ArticleCard article={mockArticle} />);
    expect(getByText('Test Article')).toBeTruthy();
    expect(getByText('2024-07-24')).toBeTruthy();
  });

  it('should display the article image', () => {
    const { getByAltText } = render(<ArticleCard article={mockArticle} />);
    const img = getByAltText('article-image Test Article');
    expect(img).toBeTruthy();
    expect(img.getAttribute('src')).toBe('test-image-url');
  });

  it('should open the dialog with article details when clicked', () => {
    const { getByText, getByRole } = render(<ArticleCard article={mockArticle} />);
    const card = getByText('Test Article');
    fireEvent.click(card);
    expect(getByRole('dialog')).toBeTruthy();
    expect(getByText('This is a test description.')).toBeTruthy();
    expect(getByText('This is the test content.')).toBeTruthy();
  });

  it('should close the dialog when the close button is clicked', async() => {
    const { getByText, getByRole, queryByRole } = render(<ArticleCard article={mockArticle} />);
    const card = getByText('Test Article');
    act(() => {
      fireEvent.click(card);
    })
    await new Promise((res) => {
      setTimeout(() => res(), 800);
    })
    const dialog = getByRole('dialog');
    expect(dialog).toBeTruthy();
    act(() => {
      fireEvent.keyDown(dialog, { key: 'Escape', code: 'Escape' });
    })

    await new Promise((res) => {
      setTimeout(() => res(), 800);
    })

    expect(queryByRole('dialog')).toBeNull();
  });
});
