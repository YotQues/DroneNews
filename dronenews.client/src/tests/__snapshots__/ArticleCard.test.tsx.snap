// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleCard renders correctly 1`] = `
<div
  className="undefined gridItem"
  onClick={[Function]}
>
  <div>
    <img
      alt="article-image Test Article"
      onError={[Function]}
      src="test-image-url"
    />
  </div>
  <div>
    <div>
      <h6
        className="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
      >
        Test Article
      </h6>
      <h6
        className="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
      >
        2024-07-24
      </h6>
    </div>
  </div>
</div>
`;
